File: lib/render/block.dart
import 'dart:convert';
import 'dart:io';

import '../util.dart';
import '../world.dart';
import 'item.dart';
import 'render.dart';

class Block extends Render {

  Item? item;

  late bool multipart;
  late List<BlockState> blockstates;

  Block(super.location, super.name, this.item, this.multipart, this.blockstates);

  Block.json(super.location, super.json) : super.json();

  Block.resource(super.location, super.name, Map<String, dynamic> json) {
    resource(json);
  }

  @override
  void json(Map<String, dynamic> json) {
    super.json(json);
    multipart = json["multipart"];
    blockstates = List.generate(json["blockstates"].length, (index) => BlockState.json(json["blockstates"][index]));
    if(json["item"] != null) {
      item = Item.json(Location.json(json["item"]), loader.resourceJson("models/item", Location.json(json["item"])));
    }
  }
  
  @override
  Map<String, dynamic> toJson() => super.toJson()..addAll({
    if(item != null) "item": item!.location,
    "multipart": multipart,
    "blockstates": List.generate(blockstates.length, (index) => blockstates[index].toJson())
  });
  
  void resource(Map<String, dynamic> json) {
    multipart = json["multipart"] != null;
    if(multipart) {
      blockstates = List.generate(json["multipart"].length, (index) => BlockState.resource(json["multipart"][index]["apply"], json["multipart"][index]["when"]));
    } else {
      blockstates = [
        for(String condition in json["variants"].keys)
          BlockState.resource(json["variants"][condition], condition)
      ];
    }
    List<int>? bytes = loader.resource("models/item", location, "json");
    if(bytes != null) {
      item = Item.resource(location, name, jsonDecode(String.fromCharCodes(bytes)));
    }
  }

  List<BlockModel> model(Map<String, dynamic> conditions) {
    if(multipart) {
      List<BlockModel> rModels = [];
      for(BlockState blockstate in blockstates) {
        if(blockstate.condition == conditions) {
          rModels.add(blockstate.models.random);
        }
      }
      return rModels;
    } else {
      for(BlockState blockstate in blockstates) {
        if(blockstate.condition == conditions) {
          return [blockstate.models.random];
        }
      }
      return [];
    }
  }

  Future<void> save() async {
    File file = loader.dataFile("block", location, "json");
    await file.create(recursive: true);
    await file.writeAsString(encoder.convert(toJson()).replaceAllMapped(
        RegExp(r'\[\s*([\d.,\s]+)\s*\]'), (match) => '[${match.group(1)!.replaceAll(RegExp(r'\s+'), ' ').trim()}]'
    ));
    if(item != null) {
      await item!.save();
    }
  }
}

class BlockState implements JsonMappable<Map<String, dynamic>> {

  late Condition condition;
  late RandomList<BlockModel> models;

  BlockState(this.models);

  BlockState.json(Map<String, dynamic> json) {
    this.json(json);
  }

  BlockState.resource(models, condition) {
    resource(models, condition);
  }

  void json(Map<String, dynamic> json) {
    condition = Condition.json(json["condition"]);
    models = RandomList(List.generate(json["models"].length, (index) => BlockModel.json(json["models"][index])));
  }

  Map<String, dynamic> toJson() => {
    "condition": condition.toJson(),
    "models": List.generate(models.list.length, (index) => models.list[index].toJson())
  };

  void resource(models, condition) {
    this.condition = Condition.resource(condition);
    this.models = models is List ?
      RandomList(List.generate(models.length, (index) => BlockModel.resource(loader.resourceJson("models", Location.json(models[index]["model"])), rotation: Rotation.json(models[index])))) :
      RandomList([BlockModel.resource(loader.resourceJson("models", Location.json(models["model"])), rotation: Rotation.json(models))]);
  }
}

class Condition implements JsonMappable<Map<String, dynamic>> {

  late Map<String, dynamic> conditions;

  Condition(this.conditions);

  Condition.json(Map<String, dynamic> json) {
    this.json(json);
  }

  Condition.resource(json) {
    resource(json);
  }

  void json(Map<String, dynamic> json) {
    conditions = json;
  }

  Map<String, dynamic> toJson() => conditions;

  void resource(json) {
    if(json is String && json.isNotEmpty) {
      conditions = {
        for(String condition in json.split(","))
          condition.substring(0, condition.indexOf("=")): condition.substring(condition.indexOf("=") +1)
      };
    } else if(json is Map<String, dynamic>) {
      conditions = json;
    } else {
      conditions = {};
    }
  }

  //TODO
  @override
  bool operator ==(Object other) {
    Map<String, dynamic> otherConditions = other is Condition ? other.conditions : other as Map<String, dynamic>;
    for(String condition in conditions.keys) {
      if(otherConditions.containsKey(condition) && otherConditions[condition].toString() != conditions[condition]) {
        return false;
      }
    }
    return true;
  }
}

class BlockModel implements JsonMappable<Map<String, dynamic>> {

  Rotation rotation = Rotation.zero;
  List<Cube> cubes = [];

  BlockModel(this.cubes);

  BlockModel.json(Map<String, dynamic> json) {
    this.json(json);
  }

  BlockModel.resource(Map<String, dynamic> json, {Rotation? rotation, Map<String, String> pTextures = const {}}) {
    this.rotation = rotation ?? Rotation.zero;
    resource(json, pTextures);
  }

  void json(Map<String, dynamic> json) {
    if(json["rotation"] != null) {
      rotation = Rotation.json(json["rotation"]);
    }
    cubes = List.generate(json["cubes"].length, (index) => Cube.json(json["cubes"][index]));
  }

  Map<String, dynamic> toJson() => {
    if(rotation != Rotation.zero) "rotation": rotation.toJson(),
    "cubes": List.generate(cubes.length, (index) => cubes[index].toJson())
};

  void resource(Map<String, dynamic> json, Map<String, String> pTextures) {
    Map<String, String> textures = {
      if(json["textures"] != null)
        for(String key in json["textures"].keys)
          key: json["textures"][key][0] == "#" ? pTextures[json["textures"][key].substring(1)] ?? json["textures"][json["textures"][key].substring(1)]! : json["textures"][key],
      for(String key in pTextures.keys)
        key: pTextures[key]!
    };
    if(json["elements"] != null) {
      cubes = List.generate(json["elements"].length, (index) => Cube.resource(json["elements"][index], textures));
    }
    if(json["parent"] != null) {
      cubes.addAll(BlockModel.resource(loader.resourceJson("models", Location.json(json["parent"])), pTextures: textures).cubes);
    }
  }
}

File: src/minecraft/objects/block.ts
import { Rotation3D } from '../../world/world3D.js'
import { MinecraftObject } from './object.js'
import { Item } from './item.js'
import { loader, Location } from '../loader.js'
import { Cube, RenderObject } from '../render.js'
import path from 'path'

export class Block extends MinecraftObject {

  item?: Item
  multipart: boolean
  blockstates: BlockState[]

  properties: Record<string, string[]>

  constructor(location: Location, name: string, item: Item | undefined, multipart: boolean, blockstates: BlockState[], properties: Record<string, string[]> = {}) {
    super(location, name)
    this.item = item
    this.multipart = multipart
    this.blockstates = blockstates
    this.properties = properties
  }

  static fromJson(location: Location, json: any): Block {
    const itemLoc = json.item ? Location.fromJson(json.item) : undefined
    return new Block(
      location, 
      json.name, 
      json.multipart, 
      json.states[0]?.addictionable === true, 
      json.states.map((blocstate: any) => BlockState.fromJson(blocstate)),
      json.properties
    )
  }

  toJson(): { [key: string]: any } {
    return {
      name: this.name,
      item: this.item?.location.toJson(),
      multipart: this.multipart,
      states: this.blockstates.map((blocstate, i) => blocstate.toJson(this.location, i)),
      properties: this.properties
    }
  }

  static resource(location: Location, name: string, json: any): Block {
    const multipart = json.multipart !== undefined
    let blockstates
    let item

    if (multipart) {
      blockstates = json.multipart.map((part: any) => BlockState.resource(part.apply, part.when))
    } else {
      blockstates = Object.keys(json.variants).map((condition) => BlockState.resource(json.variants[condition], condition))
    }
    const itemJson = loader.resourceJson('models/item', location)
    if (itemJson) {
      item = new Item(location, name, itemJson)
    }

    const block = new Block(location, name, item, multipart, blockstates)
    block.buildProperties()
    return block
  }

  defaultRender(): RenderObject {
    return this.blockstates[0].models[0].render
  }

  secondRender(): RenderObject | undefined {
    return this.item?.defaultRender()
  }

  model(conditions: Condition | Record<string, string>): [number, number][] {
    let indexes: [number, number][] = []

    if (this.multipart) {
      this.blockstates.forEach((blockstate, i) => {
        if(blockstate.condition.equals(conditions)) {
          indexes.push([i, Math.floor(Math.random() * blockstate.models.length)])
        }
      })
    } else {
      this.blockstates.forEach((blockstate, i) => {
        if(blockstate.condition.equals(conditions)) {
          indexes = [[i, Math.floor(Math.random() * blockstate.models.length)]]
        }
      })
    }

    return indexes
  }

  buildProperties() {
    this.properties = {}
    this.blockstates.forEach((blockstate) => {
      Object.entries(blockstate.condition.conditions).forEach((property) => {
        if(this.properties[property[0]] === undefined) {
          this.properties[property[0]] = [property[1]]
        } else if(!this.properties[property[0]].includes(property[1])) {
          this.properties[property[0]].push(property[1])
        }
      }) 
    })
  }

  get path(): string {
    return loader.dataFile("block", this.location, "json")
  }

  renderToSave(): Record<string, RenderObject> {
    let renders: [PropertyKey, RenderObject][] = []

    this.blockstates.forEach((blockstate, i) => {
       blockstate.models.forEach((model, j) => {
        renders.push([path.join(loader.renderDir, 'blocks', `${this.location.toDir()}-${i}-${j}.json`), model.render])
       })
    })

    return Object.fromEntries(renders)
  }
}

export class BlockState {
  
  constructor(readonly models: BlockModel[], readonly condition: Condition = Condition.all()) { }

  static resource(models: any, condition: any): BlockState {
    return new BlockState(
      Array.isArray(models)
        ? models.map((model: any) => BlockModel.resource(loader.resourceJson('models', Location.fromJson(model.model)), {}, Rotation3D.fromGradeAxis(model)))
        : [BlockModel.resource(loader.resourceJson('models', Location.fromJson(models['model'])), {}, Rotation3D.fromGradeAxis(models))],
      Condition.resource(condition)
    )
  }

  static fromJson(json: any): BlockState {
    return new BlockState(json.models.map((model: any) => new BlockModel(model)), Condition.fromJson(json.condition))
  }

  toJson(location: Location, blockstate: number): {} {
    return {
      condition: this.condition.toJson(),
      models: this.models.map((model, i) => model.toJson(location, blockstate, i))
    }
  }
}

export class Condition {

  constructor(readonly conditions: Record<string, string>) { }

  static all(): Condition {
    return new Condition({})
  }

  static fromJson(json: any): Condition {
    return new Condition(json)
  }

  static resource(json: any): Condition {
    if (typeof json === 'string' && json.length > 0) {
      return new Condition(Object.fromEntries(
        json.split(',').map(condition => {
          const [key, value] = condition.split('=')
          return [key, value]
        })
      ))
    } else if (typeof json === 'object') {
      return new Condition(json)
    } else {
      return new Condition({})
    }
  }

  toJson(): any {
    return this.conditions
  }

  equals(other: Condition | Record<string, string>): boolean {
    const conditions = other instanceof Condition ? other.conditions : other
    return Object.entries(this.conditions).every(([key, value]) => 
      conditions[key] && conditions[key].toString() === value.toString()
    )
  }
}

export class BlockModel {

  constructor(readonly render: RenderObject) { }

  static resource(json: any, pTextures: Record<string, string>, rotation?: Rotation3D): BlockModel {
    let cubes: Cube[] = []

    const textures = {
      ...(json.textures && Object.fromEntries(
        Object.keys(json.textures).map((key) => {
          const value = json.textures[key]
          return [
            key,
            (value as string)[0] === '#' ? pTextures[(value as string).substring(1)] || json.textures[(value as string).substring(1)] : value
          ]
        })
      )),
      ...pTextures
    }

    if (json.elements) {
      cubes = json.elements.map((element: any) => Cube.resource(element, textures))
    }

    if (json.parent) {
      const parentModel = BlockModel.resource(loader.resourceJson('models', Location.fromJson(json.parent)), textures)
      cubes.push(...parentModel.render.cubes)
    }

    if(rotation) {
      cubes.forEach((cube) => cube.rotate(rotation))
    }

    return new BlockModel(new RenderObject(cubes))
  }

  static fromJson(json: any): BlockModel {
    return new BlockModel(json.cubes.map((cube: any) => Cube.fromJson(cube)))
  }

  toJson(location: Location, blockstate: number, index: number): any {
    return {
      render: path.join(loader.renderDir, 'blocks', `${location.toDir()}-${blockstate}-${index}.json`)
    }
  }
}